# Map, Filter, Reduce, Lambda & Recursion

## Tasks Today:

1) <b>Lambda Functions</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Syntax <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Saving to a Variable <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) Multiple Inputs <br>
 &nbsp;&nbsp;&nbsp;&nbsp; d) Passing a Lambda into a Function <br>
 &nbsp;&nbsp;&nbsp;&nbsp; e) Returning a Lambda from a Function <br>
 &nbsp;&nbsp;&nbsp;&nbsp; f) In-Class Exercise #1 <br>
2) <b>Map</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Syntax <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Using Lambda's with Map <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #2 <br>
3) <b>Filter</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Syntax <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Using Lambda's with Filter <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #3 <br>
4) <b>Reduce</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Syntax <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Using Lambda's with Reduce <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #4 <br>
5) <b>Recursion</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Implementing a Base <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Writing a Factorial Function <br>

6) <b>Generators & Iterators</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Yield Keyword <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Inifinite Generator <br>

7) <b>Exercises</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Exercise #1 - Filtering Empty Strings <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Exercise #2 - Sorting with Last Name <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) Exercise #3 - Conversion to Farhenheit <br>
 &nbsp;&nbsp;&nbsp;&nbsp; d) Exercise #4 - Fibonacci Sequence <br>

## Lambda Functions <br>
<p>Lambda functions... or "Anonymous Functions" are referring to inline functions with no name. The keyword lambda denotes the no name function, and executes within a single line. Without saving it to a variable; however, it is not able to be used, unless passed in either as a paramater or within list comprehension.<br>Written as "(keyword lambda) (one or more inputs) (colon) (function to be executed)"</p>

#### Syntax

In [10]:
x = lambda num: (num+2) * (num+2)
print(x(9))

volume = lambda l,w,h: l*w*h
print(volume(2, 3, 4))

121
24


#### Saving to a Variable

#### Multiple Inputs

#### Passing a Lambda into a Function

#### Returning a Lambda from a Function

In [11]:
def mulitTool(func, num):
    return (func(num))

mulitTool(lambda x: x /2, 3)

1.5

#### If Statements within Lambdas

In [13]:
#true if condition else false

evens = lambda num: f'{num} is even' if  num % 2 == 0 else f'{num} is odd...'
print(evens(12))
print(evens(13))

12 is even
13 is odd...


#### In-Class Exercise #1 <br>
<p>Write an anonymous function that cubes the arguments passed in and assign the anonymous function to a variable 'f'.</p>

In [14]:
f = lambda num: num ** 3
print(f(3))

27


## Map <br>
<p>The map function allows you to iterate over an entire list while running a function on each item of the list. This is why the map function works well with lambda's, because it simplifies things and you write less lines of code.<br>The syntax for a map function is "map(function to be used, list to be used)"<br>However, you must be careful, as the map function returns a map object, not a list. To turn it into a list we use the list() type conversion.</p>

#### Syntax

In [17]:
myTrees = ['pine', 'oak', 'maple', 'live oak', 'willow']
myNums = [1, 1, 123, 14, 15, 5451, 12, 2, .74564]

mapTrees = list(map(lambda i: i+' trees are amazing!', myTrees))


mapNums = list(map(lambda num: num * 3.141592653589793, myNums))

print(mapTrees)
print(mapNums)

['pine trees are amazing!', 'oak trees are amazing!', 'maple trees are amazing!', 'live oak trees are amazing!', 'willow trees are amazing!']
[3.141592653589793, 3.141592653589793, 386.41589639154455, 43.982297150257104, 47.12388980384689, 17124.82155471796, 37.69911184307752, 6.283185307179586, 2.3424971462226933]


#### Using Lambda's with Map

#### In-Class Exercise #2 <br>
<p>Use the map function to double each number and minus it by one in the list by using a lambda function</p>

In [19]:
numnums = [4, 5, 6, 7, 78, 1123]
print(list(map(lambda num: num*2-1, numnums)))

#Parenthesis
#Exponents
#Multiplication
#Division
#Addition
#Subtraction

[7, 9, 11, 13, 155, 2245]


## Filter() <br>
<p>Filter's are similar to the map function, where you're able to pass a function argument and a list argument and filter out something from the list based on the conditions passed. Similar to the map function, it returns a filter object, so you need to type convert it to a list()</p>

#### Syntax

In [22]:
print(list(filter(lambda num: num % 2 == 0, numnums)))

[4, 6, 78]


#### Using Lambda's with Filter()

#### In-Class Exercise #3 <br>
<p>Filter out all the numbers that are below the mean of the list.<br><b>Hint: Import the 'statistics' module</b></p>

In [30]:
numnums = [1, 2, 3, 4, 5, 6, 6, 7, 8, 8, 11, 234, 1, 1, 1, 32, 34, 2, 14, 16, 14, 17, 25, 34, 123]
mean = 0
for num in numnums:
    mean += num
mean = mean / len(numnums)
print(list(filter(lambda num: num < mean, numnums )))

print(list(filter(lambda num: num < 10, numnums)))



import re

myReg = r'USA'

myABC = ['fuckyea', 'hell yea america!', '']

[1, 2, 3, 4, 5, 6, 6, 7, 8, 8, 11, 1, 1, 1, 2, 14, 16, 14, 17]
[1, 2, 3, 4, 5, 6, 6, 7, 8, 8, 1, 1, 1, 2]


## Reduce() <br>
<p>Be very careful when using this function, as of Python 3 it's been moved to the 'functools' library and no longer is a built-in function.<br>The creator of Python himself, says to just use a for loop instead.</p>

#### Syntax

In [33]:
from functools import reduce
listy = [1, 2, 3, 4, 5, 6, 7, 8]

print(reduce(lambda x,y: x + y, listy))

36


#### Using Lambda's with Reduce()

#### In-Class Exercise #4 <br>
<p>Use the reduce function to multiply the numbers in the list below together with a lambda function.</p>

In [34]:
num3 = [1, 2, 3, 4, 5,]

print(reduce(lambda x,y: x * y, num3))


120


## Recursion <br>
<p>Recursion means that a function is calling itself, so it contanstly executes until a base case is reached. It will then push the returning values back up the chain until the function is complete. A prime example of recursion is computing factorials... such that 5! (factorial) is 5*4*3*2*1 which equals 120.</p>

#### Implementing a Base Case

In [45]:
def subtracting(num):
    if num == 1:
        return num
    else:
        print(f'subtracting({num}) = {num} - subtracting({num} - 1))')
        return num - (subtracting(num-1))
    
subtracting(5)

mylist = ['a', 'b', 'c', [1, 2, 3,['abc', 'def', 'hij']]]

def unpack(l):
    for item in mylist:
        if isinstance(item, list):
            unpack(item)
        else:
            print(item)
unpack(mylist)

subtracting(5) = 5 - subtracting(5 - 1))
subtracting(4) = 4 - subtracting(4 - 1))
subtracting(3) = 3 - subtracting(3 - 1))
subtracting(2) = 2 - subtracting(2 - 1))
a
b
c
a
b
c
a
b
c
a
b
c
a
b
c
a
b
c
a
b
c
a
b
c
a
b
c
a
b
c
a
b
c
a
b
c
a
b
c
a
b
c
a
b
c
a
b
c
a
b
c
a
b
c
a
b
c
a
b
c
a
b
c
a
b
c
a
b
c
a
b
c
a
b
c
a
b
c
a
b
c
a
b
c
a
b
c
a
b
c
a
b
c
a
b
c
a
b
c
a
b
c
a
b
c
a
b
c
a
b
c
a
b
c
a
b
c
a
b
c
a
b
c
a
b
c
a
b
c
a
b
c
a
b
c
a
b
c
a
b
c
a
b
c
a
b
c
a
b
c
a
b
c
a
b
c
a
b
c
a
b
c
a
b
c
a
b
c
a
b
c
a
b
c
a
b
c
a
b
c
a
b
c
a
b
c
a
b
c
a
b
c
a
b
c
a
b
c
a
b
c
a
b
c
a
b
c
a
b
c
a
b
c
a
b
c
a
b
c
a
b
c
a
b
c
a
b
c
a
b
c
a
b
c
a
b
c
a
b
c
a
b
c
a
b
c
a
b
c
a
b
c
a
b
c
a
b
c
a
b
c
a
b
c
a
b
c
a
b
c
a
b
c
a
b
c
a
b
c
a
b
c
a
b
c
a
b
c
a
b
c
a
b
c
a
b
c
a
b
c
a
b
c
a
b
c
a
b
c
a
b
c
a
b
c
a
b
c
a
b
c
a
b
c
a
b
c
a
b
c
a
b
c
a
b
c
a
b
c
a
b
c
a
b
c
a
b
c
a
b
c
a
b
c
a
b
c
a
b
c
a
b
c
a
b
c
a
b
c
a
b
c
a
b
c
a
b
c
a
b
c
a
b
c
a
b
c
a
b
c
a
b
c
a
b
c
a
b
c
a
b
c
a
b
c
a
b
c
a
b
c
a
b
c
a
b
c
a


RecursionError: maximum recursion depth exceeded while calling a Python object

#### Writing a Factorial Function

## Generators <br>
<p>Generators are a type of iterable, like lists or tuples. They do not allow indexing, but they can still be iterated through with for loops. They are created using functions and the yield statement.</p>

#### Yield Keyword <br>
<p>The yield keyword denotes a generator, it doesn't return so it won't leave the function and reset all variables in the function scope, instead it yields the number back to the caller.</p>

In [57]:
def myGen():
    num = 0
    while True:
        yield num
        num += 1

gen = myGen()
print(next(myGen()))
print(next(myGen()))
print(next(gen))
print(next(gen))
print(next(gen))


0
0
0
1
2


# Exercises

### Exercise #1 <br>
<p>Filter out all of the empty strings from the list below</p>

`Output: ['Argentina', 'San Diego', 'Boston', 'New York']`

In [62]:
places = [" ","Argentina", " ", "San Diego","","  ","","Boston","New York"]
print(list(filter(lambda word: word.strip() != "", places)))

['Argentina', 'San Diego', 'Boston', 'New York']


### Exercise #2 <br>
<p>Write an anonymous function that sorts this list by the last name...<br><b>Hint: Use the ".sort()" method and access the key"</b></p>

`Output: ['Victor aNisimov', 'Gary A.J. Bernstein', 'Joel Carter', 'Andrew P. Garfield', 'David hassELHOFF']`

In [74]:
author = ["Joel Carter", "Victor aNisimov", "Andrew P. Garfield","David hassELHOFF","Gary A.J. Bernstein"]

somewhatSorted = list(map(lambda name: name.split()[-1], author))
print(somewhatSorted)

fullySorted = sorted(somewhatSorted, key = lambda name: name.lower())
print(fullySorted)


# look into .sort() keys to find out how to sort by last name

#

['Carter', 'aNisimov', 'Garfield', 'hassELHOFF', 'Bernstein']
['aNisimov', 'Bernstein', 'Carter', 'Garfield', 'hassELHOFF']


### Exercise #3 <br>
<p>Convert the list below from Celsius to Farhenheit, using the map function with a lambda...</p>

`Output: [('Nashua', 89.6), ('Boston', 53.6), ('Los Angelos', 111.2), ('Miami', 84.2)]
`

In [82]:
# F = (9/5)*C + 32
# keep city names

places = [('Nashua',32),("Boston",12),("Los Angelos",44),("Miami",29)]

print(list(map(lambda x: ((9/5)*x + 32) if type(x) == int else x, places)))

[('Nashua', 32), ('Boston', 12), ('Los Angelos', 44), ('Miami', 29)]


### Exercise #4 <br>
<p>Write a recursion function to perform the fibonacci sequence up to the number passed in.</p>

`Output for fib(5) => 
Iteration 0: 1
Iteration 1: 1
Iteration 2: 2
Iteration 3: 3
Iteration 4: 5
Iteration 5: 8`

In [170]:
#0   == 1        --> 0
#1+0 == 1        --> 1
#1+1 == 2        --> 2
#2+1 == 3        --> 3
#3+2 == 5        --> 4
#5+3 == 8        --> 5
#8+5 == 13       --> 6

def fib(num):
    if num < 0:
        return 0
    if num == 0 or num == 1:
        return 1
    if num > 1:
        # print((fib(num-1) + fib(num-2))) -- doesn't work like i expected... gave up after i don't know how long...
        return fib(num-1) + fib(num-2) # - does get the right number though!
fib(6)

13